### PR TITLE
IGNITE-6333 Added 2 metrics: rollbacks due to timeouts and deadlocks

### PR DESCRIPTION
**Pull request by xtern**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java [hunk 1]
```diff
@@ -3238,6 +3241,12 @@ public void commit() throws IgniteCheckedException {
                 catch (IgniteCheckedException e) {
                     COMMIT_ERR_UPD.compareAndSet(GridNearTxLocal.this, null, e);
 
+                    if (X.hasCause(e, TransactionDeadlockException.class))
+                        GridNearTxLocal.this.deadlocked = true;
+                    else
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java]: "else if", plz no "else \n if"

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java [hunk 2]
```diff
@@ -1073,12 +1076,12 @@ protected final boolean state(TransactionState state, boolean timedOut) {
                 }
             }
 
+            if (timedOut)
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java]: why?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java [hunk 3]
```diff
@@ -221,7 +221,10 @@
     private volatile TransactionState state = ACTIVE;
 
     /** Timed out flag. */
-    private volatile boolean timedOut;
+    protected volatile boolean timedOut;
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java]: may be leave private and add protected getter

Comment (xtern) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java]: and setter

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java [hunk 4]
```diff
@@ -3238,6 +3241,12 @@ public void commit() throws IgniteCheckedException {
                 catch (IgniteCheckedException e) {
                     COMMIT_ERR_UPD.compareAndSet(GridNearTxLocal.this, null, e);
 
+                    if (X.hasCause(e, TransactionDeadlockException.class))
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java]: may be set flag in the very place where exception is thrown

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java [hunk 5]
```diff
@@ -221,7 +221,10 @@
...
+
+    /** Transaction participated deadlock flag. */
+    protected volatile boolean deadlocked;
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxAdapter.java]: may be private and create public setter

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java [hunk 6]
```diff
@@ -3238,6 +3241,12 @@ public void commit() throws IgniteCheckedException {
...
+                        GridNearTxLocal.this.deadlocked = true;
```
Comment (voipp) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxLocal.java]: may be set flag in the very place where exception is thrown

**Pull request closed without merging**